Coral/interactive.js

```diff
@@ -0,0 +1,69 @@
+export const SELECT_RADIUS = 8;
+const SELECT_RADIUS_SQR = SELECT_RADIUS * SELECT_RADIUS;
+const MAX_TANGENT_MAGNITUDE = 0.5;
+
+export class InteractiveVertex {
+  constructor(control_point, constraint, tile_quad) {
+    this.control_point = control_point;
+    this.constraint = constraint;
+    this.tile_quad = tile_quad;
+  }
+
+  get position_world() {
+    return this.tile_quad.uv_to_world(this.control_point.position);
+  }
+
+  is_hovering(mouse) {
+    const position_world = this.tile_quad.uv_to_world(
+      this.control_point.position
+    );
+    const dist_sqr = mouse.dist_sqr(position_world);
+
+    return dist_sqr < SELECT_RADIUS_SQR;
+  }
+
+  move(mouse) {
+    const uv = this.tile_quad.world_to_uv(mouse);
+    this.control_point.position = this.constraint.clamp(uv);
+  }
+}
+
+export class InteractiveTangent {
+  constructor(control_point, constraint, tile_quad) {
+    this.control_point = control_point;
+    this.constraint = constraint;
+    this.tile_quad = tile_quad;
+  }
+
+  get position_world() {
+    return this.tile_quad.uv_to_world(this.control_point.forward_point);
+  }
+
+  is_hovering(mouse) {
+    const position_world = this.tile_quad.uv_to_world(
+      this.control_point.forward_point
+    );
+    const dist_sqr = mouse.dist_sqr(position_world);
+
+    return dist_sqr < SELECT_RADIUS_SQR;
+  }
+
+  move(mouse) {
+    const uv = this.tile_quad.world_to_uv(mouse);
+    const original_tangent = uv.sub(this.control_point.position);
+
+    // If the tangent is facing backwards, negate it to flip it forwards.
```

Comment:
This was the wrong approach, I actually want to reflect it in the plane orthogonal to the constraint direction. The new logic should be:

```
if original_tangent dot constraint < 0:
  // dual computes the line, to_reflection() reinterprets it as a versor
  // This could be precomputed in the constructor
  reflection = constraint.dual().to_reflection()
  corrected_tangent = line.to_reflection().transform(original_tangent);
```